Add function to fetch and load Bedmap2

santisoler opened the pull request:
Bedmap2 is a suite of gridded products describing surface elevation,
ice-thickness, the sea ﬂoor and subglacial bed elevation of the Antarctic south
of 60°S. The Bedmap2 datasets are downloaded from the
[British Antarctic Survey](https://www.bas.ac.uk/) and loaded as `xarray.Dataset` objects.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to the docstring, gallery, and/or tutorials.

santisoler committed:
Start adding fetch_bedmap2()

santisoler committed:
Improve bedmap2 docstring

santisoler committed:
Fix datasets' names

santisoler committed:
Fix dataarray merge

santisoler committed:
Add rasterio to environment.yml

santisoler committed:
Extend fetch_bedmap2 docstring

santisoler committed:
Remove double space in docstring

santisoler committed:
Replace no data values with nans

santisoler committed:
Improve bedmap2 docstring and add it to doc index

santisoler committed:
Improve styling

santisoler committed:
Improve datasets argument

Now we can pass a str or a list of datasets as argument.
Besides, the "all" str makes every dataset to be loaded.

santisoler committed:
Improve docstring styling

santisoler committed:
Remove band dimensions from bedmap2 arrays

santisoler committed:
Make datasets argument easier

santisoler committed:
Add "all" option for datasets argument

santisoler committed:
Replace IOError for ValueError

santisoler committed:
No default datasets

The "all" argument for datasets could freeze low memory systems.

santisoler committed:
Improve dataset merging

Instead of using a list (non efficient) we merge the dataset when
loading a new array.

santisoler committed:
Add bedmap2 tests functions

santisoler committed:
Fix styling and format

santisoler commented:
@leouieda  would you mind checking this code?

I have one minor doubt: Should import the `DATASETS` list in the `__init__.py` in order to make it accesible to users?
So, if one want to load every dataset could do something like:
```python
from rockhound import fetch_bedmap2, BEDMAP2_DATASETS

grid = fetch_bedmap2(BEDMAP2_DATASETS)
```

**Note: loading every dataset could eat a lot of memory!**

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+    Bedmap2 is a suite of gridded products describing surface elevation,
+    ice-thickness, the sea ﬂoor and subglacial bed elevation of the Antarctic south
+    of 60°S [BEDMAP2]_.
+    The datasets are downloaded as `tiff` files and loaded into a
```
```suggestion
    The datasets are downloaded as ``tiff`` files and loaded into a
```

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+    - ``rockmask``: mask showing rock outcrops
+    - ``lakemask_vostok``: mask showing the extent of the lake cavity of Lake Vostok
+    - ``grounded_bed_uncertainty``: ice bed uncertainty grid
+      (figure 12 of the manuscript)
```
```suggestion
```

Either that or include references to all other grids. I don't think we need this because it should be plotted in our gallery example.

santisoler replied:
You're right. I've taken the datasets description from the `bedmap2.readme` file.
I'll remove the `(figure n of the manuscript)` pieces on every appearance.

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+    Parameters
+    ----------
+    datasets : list
+        Datasets that will be loaded from the Bedmap2 model.
```
```suggestion
        Names of the datasets that will be loaded from the Bedmap2 model.
```

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+        Datasets that will be loaded from the Bedmap2 model.
+    load : bool
+        Wether to load the data into an :class:`xarray.Dataset` or just return the
+        path to the downloaded data.
```
```suggestion
        path to the downloaded data in a tiff file. If False, the *datasets* argument is ignored.
```

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+        return fname
+
+    for dataset in datasets:
+        if dataset not in DATASETS:
```
This could be cleaner using sets:

```python
if not set(datasets).issubset(DATASETS):
    raise ValueError("Invalid datasets: {}".format(set(datasets).difference(DATASETS)))
```

And we can catch all invalid values in one go.

santisoler replied:
Nice!

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+
+    Each dataset is projected in Antarctic Polar Stereographic projection, latitude of
+    true scale -71 degrees south, datum WGS84. All heights are in metres relative to
+    sea level as defined by the g104c geoid.
```
We should probably add the projection and datum information to the `Dataset.attrs`

santisoler replied:
Done!

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+            raise ValueError("Dataset {} not found in bedmap2_tiff.zip".format(dataset))
+    available_datasets = dict(
+        zip(DATASETS, ["bedmap2_{}.tif".format(dataset) for dataset in DATASETS])
+    )
```
The `zip` method is great if you already have two lists. But if you're doing a comprehension, then a dict comprehension is more convenient:

```python
available_datasets = {dataset: "bedmap2_{}.tif".format(dataset) for dataset in DATASETS}
```

santisoler replied:
That's better, I'll change it.

leouieda commented on environment.yml:
```diff
@@ -8,6 +8,7 @@ dependencies:
     - pooch
     - xarray
     - pandas
+    - rasterio
```
This needs to be added to `requirements.txt`, `setup.py`, and `doc/install.rst`.

santisoler replied:
Done!

leouieda commented on rockhound/tests/test_bedmap2.py:
```diff
@@ -0,0 +1,53 @@
+        "geoid": [np.float32(-65.86805), np.float32(36.63612)],
+    }
+    for dataset in datasets:
+        grid = fetch_bedmap2(datasets=datasets)
```
```suggestion
        grid = fetch_bedmap2(datasets=[dataset])
```

santisoler replied:
Thanks for that! My mistake!

santisoler committed:
Change import order of numpy and pytest

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+        zip(DATASETS, ["bedmap2_{}.tif".format(dataset) for dataset in DATASETS])
+    )
+    available_datasets["geoid"] = "gl04c_geiod_to_WGS84.tif"
+    for i, dataset in enumerate(datasets):
```
```suggestion
    grids = []
    for dataset in datasets:
```

santisoler committed:
Add rasterio to setup.py as install requirement

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,103 @@
+            if i == 0:
+                grid = array.to_dataset()
+            else:
+                grid = xr.merge([grid, array])
```
```python
    grids.append(array)
grid = xr.merge(grids)
```

leouieda replied:
The metadata should be set after the merge because xarray doesn't carry the `attrs` over to the new grid.

santisoler replied:
I thought that using lists would eat more memory, but I've just checked that it's the same, so I'll change the code according to your proposal (i.e. like I've implemented it on early stages).

leouieda reviewed (changes requested):
@santisoler looks great! Just a few suggestions for making the code a bit cleaner. 

I don't think we should expose the `DATASETS` variable. That's a very narrow use case. What we could do is allow `datasets="all"` to be a valid input. But we can't load some of the grids into the same `Dataset` because they have different coordinates and shapes, right?

santisoler commented:
> @santisoler looks great! Just a few suggestions for making the code a bit cleaner.
> 
> I don't think we should expose the `DATASETS` variable. That's a very narrow use case. What we could do is allow `datasets="all"` to be a valid input. 

I've been messing around with the `datasets="all"` argument with the idea of making it the default value, but due to the high memory requirement I dropped it. But leaving the `datasets="all"` as an option would be good. I'll add it.

> But we can't load some of the grids into the same `Dataset` because they have different coordinates and shapes, right?

Thanks to `xarray` we could get grids with different shapes on the same `Dataset` with the `xarray.merge()` function. The coordinates are spanned to include the coordinates of every `DataArray` and the values are filled with nans if no data is available.
Don't know if we should only merge `DataArray`s with the same shape or leave the code as it is now.

santisoler commented:
@leouieda I'll apply the requested changes tomorrow. Thanks for the review!

leouieda commented:
> Don't know if we should only merge `DataArray`s with the same shape or leave the code as it is now.

That's a tough one. Maybe we can leave as is but include a warning the docstring that you'll get different things if you ask for only the uncertainty grids or for all of them.

leouieda commented:
> due to the high memory requirement I dropped it

This should definitely be a warning in the docstring. We don't want people crashing their computers. And it should definitely not be the default :smiley:

leouieda committed:
Minor fix in docstring

Co-Authored-By: santisoler <santiago.r.soler@gmail.com>

leouieda committed:
Improve docstring

Co-Authored-By: santisoler <santiago.r.soler@gmail.com>

leouieda committed:
Improve docstring

Co-Authored-By: santisoler <santiago.r.soler@gmail.com>

santisoler committed:
Remove references to figures in manuscript

santisoler committed:
Improve check for invalid datasets

santisoler committed:
Improve build of available_datasets dict

leouieda committed:
Fix wrong datasets argument in test

Co-Authored-By: santisoler <santiago.r.soler@gmail.com>

santisoler committed:
Add rasterio as dependency

santisoler committed:
Fix style in docstring

santisoler committed:
Add option to pass datasets as strings

Also add the dataset="all" option to load every dataset and the
corresponding test function.

santisoler committed:
Add high memory usage warning

santisoler committed:
Add shape alteration warning

santisoler committed:
Add projection data to grid attrs

santisoler committed:
Improve code style when merging arrays

santisoler commented:
@leouieda I think the code and its docstring look better now.
I've noticed that the only the dataset `thickness_uncertainty_5km` is defined on a different set of points than the rest, so we don't have to bother with the `lakemask_vostok` one.
That's why the warning only reference the former as source of problems.

Do you like the hard-coded metadata as it is?

santisoler committed:
Improve code style to fix lint errors

leouieda commented on rockhound/tests/test_bedmap2.py:
```diff
@@ -19,6 +20,16 @@ def test_bedmap2_file_name_only():
+        grid = fetch_bedmap2(dataset)
+        assert set(grid.data_vars) == set([dataset])
+    # Test docstrings="all"
+    grid = fetch_bedmap2("all")
```
This seems kind of dangerous since you said that it would eat a lot of memory.

santisoler replied:
Yes. It could be a source of lot of headaches in the future.

leouieda commented on rockhound/bedmap2.py:
```diff
@@ -0,0 +1,122 @@
+
+    .. warning ::
+        Loading a great number of datasets may require a fair amount of memory that
+        could crash your system. We recommend loading only the needed datasets.
```
Now that I read this docstring it seems kind of contradictory to say this and then offer an `all` option. It might be better to remove the `all`. It'll make the code simpler and we don't have to test loading all the data, which will certainly crash the CI.

leouieda commented:
:+1: I really like the changes and the docstring! Good job!

What do you think about removing the `all` option? It feels like something that will be abused and cause system crashes often. Plus, we can't test it on the CIs.

santisoler commented:
@leouieda Thanks for the reviews.

I think I've added and removed the `all` option a hundred times, what leads me to think we are not entirely sure if it worth the risk of crashing local systems and/or Travis (is it possible?).
We definitely should remove it but leave the option to pass the datasets are strings (instead of putting a single dataset inside a list). In fact I don't imagine a case in which someone would need the entire set.

I'll remove it from the code, the docstring and the test functions. Then if no CI fails we could merge.

santisoler committed:
Remove "all" option

santisoler committed:
Add test function for multiple datasets

santisoler committed:
Fix datasets iteration

leouieda commented:
:+1: thanks, Santi!

leouieda commented:
We still need a gallery plot for this. Given the size of the grids, probably only plot a subsection, like we did for the ETOPO1 grid.

santisoler commented:
Thanks for the reviews @leouieda! I can get into the gallery plot this week. I'll open a PR when ready.
